### PR TITLE
Referral codes

### PR DESCRIPTION
**Pull request by jkeat**
Links
-----
- Web repo PR: https://github.com/UPchieve/web/pull/382

Description
-----------
- Generate referral code based on user ID
- Script for adding referral codes to existing users
- Store referrer on signup

Developer self-review checklist
-------------------------------
- [ ] Potentially confusing code has been explained with comments
- [ ] No warnings or errors have been introduced; all known error cases have been handled
- [ ] Any appropriate documentation (within the code, README.md, docs, etc) has been updated
- [ ] All edge cases have been addressed

### CODE REVIEW

File: dbutils/add-referral-codes.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+const mongoose = require('mongoose')
+const base64url = require('base64url')
+const dbconnect = require('./dbconnect')
+const User = require('../models/User')
+
+dbconnect(mongoose, function() {
+  User.find()
```
Comment (ericyliu) [inline dbutils/add-referral-codes.js:7]: consider using async/await here?

File: router/auth/index.js [hunk 2]
```diff
@@ -103,7 +104,7 @@ module.exports = function(app) {
     })
   })
 
-  router.post('/register', function(req, res, next) {
+  router.post('/register', async function(req, res, next) {
```
Comment (ericyliu) [inline router/auth/index.js:107]: this is a pretty gnarly routing function

generally, its cleaner to abstract away the business logic of updating the db to a service layer and have the routing layer call functions in that layer

Comment (jkeat) [reply router/auth/index.js:107]: so gnarly. definitely want this to get refactored soon

File: dbutils/add-referral-codes.js (hunk 1 above)
Comment (jkeat) [reply dbutils/add-referral-codes.js:7]: Just gonna leave it as is b/c it's a one-off script, but definitely more async/await usage going forward 👍

**Pull request merged**